src/main/java/appeng/parts/misc/ItemHandlerAdapter.java

```diff
@@ -0,0 +1,399 @@
+					if( gathered == null )
+					{
+						gathered = extracted;
+					}
+					else
+					{
+						gathered.stackSize += extracted.stackSize;
+					}
+					remainingSize -= gathered.stackSize;
+				}
+			}
+			while( extracted != null && remainingSize > 0 );
+
+			// Done?
+			if( remainingSize <= 0 )
+			{
+				break;
+			}
+		}
+
+		if( gathered != null )
+		{
+			if( mode == Actionable.MODULATE )
+			{
+				this.onTick();
+			}
+
+			return AEItemStack.create( gathered );
+		}
+
+		return null;
+	}
+
+	@Override
+	public TickRateModulation onTick()
+	{
+		LinkedList<IAEItemStack> changes = new LinkedList<>();
+
+		int slots = itemHandler.getSlots();
+
+		// Make room for new slots
+		if( slots > cachedStacks.length )
+		{
+			cachedStacks = Arrays.copyOf( cachedStacks, slots );
+			cachedAeStacks = Arrays.copyOf( cachedAeStacks, slots );
+		}
+
+		for( int slot = 0; slot < slots; slot++ )
+		{
+			// Save the old stuff
+			ItemStack oldIS = cachedStacks[slot];
+			IAEItemStack oldAeIS = cachedAeStacks[slot];
+
+			ItemStack newIS = itemHandler.getStackInSlot( slot );
+
+			if( this.isDifferent( newIS, oldIS ) )
+			{
+				addItemChange( slot, oldAeIS, newIS, changes );
+			}
+			else if( newIS != null && oldIS != null )
+			{
+				addPossibleStackSizeChange( slot, oldAeIS, newIS, changes );
+			}
+		}
+
+		// Handle cases where the number of slots actually is lower now than before
+		if( slots < cachedStacks.length )
+		{
+			for( int slot = slots; slot < cachedStacks.length; slot++ )
+			{
+				IAEItemStack aeStack = cachedAeStacks[slot];
+				if( aeStack != null )
+				{
+					IAEItemStack a = aeStack.copy();
+					a.setStackSize( -a.getStackSize() );
+					changes.add( a );
+				}
+			}
+
+			// Reduce the cache size
+			cachedStacks = Arrays.copyOf( cachedStacks, slots );
+			cachedAeStacks = Arrays.copyOf( cachedAeStacks, slots );
+		}
+
+		if( !changes.isEmpty() )
+		{
+			this.postDifference( changes );
+			return TickRateModulation.URGENT;
+		}
+		else
+		{
+			return TickRateModulation.SLOWER;
+		}
+	}
+
+	private void addItemChange( int slot, IAEItemStack oldAeIS, ItemStack newIS, List<IAEItemStack> changes )
+	{
+		// Completely different item
+		cachedStacks[slot] = newIS;
+		cachedAeStacks[slot] = AEItemStack.create( newIS );
+
+		// If we had a stack previously in this slot, notify the newtork about its disappearance
+		if( oldAeIS != null )
+		{
+			oldAeIS.setStackSize( -oldAeIS.getStackSize() );
+			changes.add( oldAeIS );
+		}
+
+		// Notify the network about the new stack. Note that this is null if newIS was null
+		if( cachedAeStacks[slot] != null )
+		{
+			changes.add( cachedAeStacks[slot] );
+		}
+	}
+
+	private void addPossibleStackSizeChange( int slot, IAEItemStack oldAeIS, ItemStack newIS, List<IAEItemStack> changes )
+	{
+		// Still the same item, but amount might have changed
+		long diff = newIS.stackSize - oldAeIS.getStackSize();
+
+		if( diff != 0 )
+		{
+			IAEItemStack stack = oldAeIS.copy();
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'stack' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.orod.org/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)